state-sync/state-sync-v2/data-streaming-service/src/data_stream.rs

```diff
@@ -412,23 +409,53 @@ impl<T: AptosDataClientInterface + Send + Clone + 'static> DataStream<T> {
         // Process any ready data responses
         for _ in 0..self.get_max_concurrent_requests() {
             if let Some(pending_response) = self.pop_pending_response_queue()? {
+                // Get the client request and response information
                 let maybe_client_response = pending_response.lock().client_response.take();
                 let client_response = maybe_client_response.ok_or_else(|| {
                     Error::UnexpectedErrorEncountered("The client response should be ready!".into())
                 })?;
                 let client_request = &pending_response.lock().client_request.clone();
 
+                // Process the client response
                 match client_response {
                     Ok(client_response) => {
-                        if sanity_check_client_response(client_request, &client_response) {
+                        // Sanity check and process the response
+                        if sanity_check_client_response_type(client_request, &client_response) {
+                            // If the response wasn't enough to satisfy the original request (e.g.,
+                            // it was truncated), missing data should be requested.
+                            let missing_data_requested = self.missing_data_requested(
```

Comment:
Should the function name be more active like `request_missing_data()`? The current name sounds like a simple check rather than actually requesting things

Reply:
Maybe a nit: is there any reason to request missing data before notifying the client of the data? It seems a bit easier to reason if all the data request logic is in one place instead of surrounding the data notification. E.g., I'm not sure if the error handling is correct if we return an error for the data notification but in the meantime have already requested missing data.

Reply:
> Should the function name be more active like request_missing_data()? 

Yeah, that make sense 😄 Updated it to `request_missing_data` and updated the method's comment.

> is there any reason to request missing data before notifying the client of the data?

Aah, yes! I did it to avoid having to clone the response. If you look at the current implementation, we just use references in `request_missing_data` and then `send_data_notification_to_client` takes the actual response. When I flipped it around, it requires me to clone 😢

Let me know what you think. I agree that it would be cleaner to send it first. To be fair, the clone probably isn't that bad 😄